homeassistant/helpers/entity.py

```diff
@@ -489,6 +489,8 @@ def _stringify_state(self) -> str:
             # If the entity's state is a float, limit precision according to machine
             # epsilon to make the string representation readable
             return f"{state:.{FLOAT_PRECISION}}"
+        if isinstance(state, datetime):
```

Comment:
I think each entity integration, eg sensor, should take care of any conversion needed. The integration better knows what kind of state is expected and allowed.

Reply:
I don't get this comment. We added support for handling temperature conversions, accepts floats and ints... we all convert it to a string in the state machine. How is this different?

The reason I put this in is to reduce issues with incorrectly converted datetime. As a matter of fact, you could return a DateTime object right now and it would work... except not as good in the frontend.

Returning a timestamp is a common thing, let's support it and make sure we always handle it correctly and never have to code review comments about using `isoformat()` again.

Reply:
Each integration should have defined return values. A sensor datetime sensor should eg have a datetime return value type if we decide that. Then the sensor integration converts it to a string. There will be no guessing or instance type checking.

This is why we moved the temperature conversion to the sensor integration and introduced native value instead of extending the entity helper.

Reply:
Right, so moving that logic into the sensor instead of here. got it 👍

Reply:
If possible I think we should check for device class instead of instance type and that way enforce the type for the device class. Then we'll error early and hard, when converting to isoformatted string for datetime and date device class.

I think it's a great idea to move the conversion away from the platforms. 👍

Reply:
Isn't that the same as I purposed here? <https://github.com/home-assistant/core/pull/56100>

Reply:
It's similar but I don't think we should catch the error. We should blow up. That way faulty integrations will be caught and fixed immediately.

Reply:
- Moved it to the sensor platform
- Blow up in case it is invalid

Thanks for that reference @ludeeus, picked up that part in this PR as well.